### ISSUE TITLE
Make recognized file extensions configurable

### ISSUE THREAD

**Issue by oliver-sc** [CJCombrink/SpellChecker-Plugin#95]
In our project we use the file extension ".doxy" for source files which just contain comments in doxygen syntax (e.g. for a chapter "Getting Started" etc.). First, I tried to get this spell checked by adding "*.doxy" as MIME type for c++ sources. This didn't work.
After looking closer into the SpellChecker source code I realized that ".dox" is the file extension intended for this. Are there plans to make the list of file extensions configurable for the user or take the list from the MIME type settings as proposed in issue #39?
BTW: Why is this checked needed at all?

**Comment by CJCombrink** [comment]
@oliver-sc 
It is definitely something I still want to do, but did not get the time yet to do so. 
I had a similar issue with one of the projects I worked on and there the extension was dox, thus the reason for adding it. 

The reason why the extension is checked is that the code forms part of the C++ parser, I got into situations where the C++ parser tried to parse non-C++ files and that just didn't work, or in some cases even crashed the parser. At that time this was the only solution I could think of, I was not aware of the MIME database at that time.

**Comment by oliver-sc** [comment]
@CJCombrink 

> The reason why the extension is checked is that the code forms part of the C++ parser, I got into situations where the C++ parser tried to parse non-C++ files and that just didn't work, or in some cases even crashed the parser. At that time this was the only solution I could think of, I was not aware of the MIME database at that time.

Ok, I understand. Although I think the parser should never crash if it parses a malformed file. Such situation can also happen if you open a *.cpp file with some syntax errors (of course much more unlikely).

Regarding using the MIME database:
After looking at the MIME types, I'm not fully convinced that this is the best solution. There are several MIME types that map to the C++ Editor, e.g.
```
C/C++:
text/x-csrc    => C++ Editor
text/x-c++src  => C++ Editor
text/x-chdr    => C++ Editor
text/x-c++hdr  => C++ Editor
Others:
text/x-csharp  => C++ Editor
text/x-objc++  => C++ Editor
etc.
```
Which one would you support? text/x-csrc, text/x-c++src, text/x-chdr, text/x-c++hdr ? What about x-objc++? You can be sure that someone would request to change/modify this list. 
I think it would be easier and more flexible to just add a file patterns entry (similar to the MIME Types pattern) in the Spell Checker Options settings. This way every user has full control over it.

**Comment by CJCombrink** [comment]
@oliver-sc 

I have started looking into the MIME Database for parsing C++'like files.
I have created pre-release for Windows and Qt Creator 4.8-beta 1. If you can test this version it will be great. If not, please let me know what is your setup and perhaps I might be able to build a test build for you.

The idea is that the C++ parser should be able to parse any file that opens with the C++ editor. 
And to allow doxygen files (*.dox, *.doxy, etc) files I have added a new MIME type: `text/x-c++dox`
If you need to change which files must be checked, the MIME database can be edited/extended.

I will continue testing (thus the specific pre-release dll) but am planning a few changes during the next few days..

Any feedback will be appreciated.

**Comment by oliver-sc** [comment]
@CJCombrink 

Just yesterday I renamed all the doxygen files in our project from *.doxy => *.dox after learning that the QtCreator syntax highlighting expects *.dox and in the doxygen configuration file under FILE_PATTERNS it says
```
# If left blank the following patterns are tested:*.c, *.cc, *.cxx, *.cpp,
# *.c++, *.java, *.ii, *.ixx, *.ipp, *.i++, *.inl, *.idl, *.ddl, *.odl, *.h,
# *.hh, *.hxx, *.hpp, *.h++, *.cs, *.d, *.php, *.php4, *.php5, *.phtml, *.inc,
# *.m, *.markdown, *.md, *.mm, *.dox, *.py, *.pyw, *.f90, *.f, *.for, *.tcl,
# *.vhd, *.vhdl, *.ucf, *.qsf, *.as and *.js.
```
i.e. *.dox seems to be the recommended file extension.

I'm working under Linux, but I will try to build your pre-release. Does it require Qt Creator 4.8-beta or should it build with 4.6.1 as well?

P.S. I will be on vacation until end of this week so it's probably next week when I can test this.

**Comment by CJCombrink** [comment]
Interesting, thanks for the feedback. 

> ... doxygen configuration file under FILE_PATTERNS ...

I will look into that as well, perhaps there is a way to also check those files even if there is not a valid editor for them. 

>  Does it require Qt Creator 4.8-beta or should it build with 4.6.1 as well?

I just had my setup for Qt Creator 4.8 beta due to other tests, and I lost the builds for older versions. If I had more time I would have built at least for 4.7 since there were no API Changes from 4.7 to 4.8 beta. 
I can not remember if there were API changes from 4.6 to 4.7, thus I am not sure if it will work without source changes. 

> P.S. I will be on vacation until end of this week so it's probably next week when I can test this.

I will probably have Linux builds by then, for at least 4.7 and 4.8-(whatever is the latest at that time). I can not guarantee older since I decided at the start that it takes too much time to support older versions compared to the gains.

**Comment by oliver-sc** [comment]
Quick update:
I managed to build the plugin for Qt Creator 4.8 beta (I had a few strange linker errors which are probably the result of different Windows vs. Linux compiler behavior). I did a few basic tests with *.dox/*.doxy files. So far things look good!

**Comment by CJCombrink** [comment]
@oliver-sc 
Thanks for the update.
Sorry I never got to building any more test releases. 

You might have seen that I have started a complete refactoring of the C++ document parser and are working on that as much as I can. 

I have the following goals (not related to the issue but maybe nice to know):
1. Strip out the common parts that can be re-used when adding a QML parser (like splitting words and applying settings, yes the QML parser is planned to come soon)
1. Better logical flow of steps so that I can better leverage parallel work and to increase maintainability.
1. Better memory management. I saw a few crashes when loading large projects due to to the memory usage of QtCreator going past the 1.2GB limit (32 bit application). There are a few options that I have investigated (like [/LARGEADDRESSAWARE](https://docs.microsoft.com/en-za/cpp/build/reference/largeaddressaware-handle-large-addresses?view=vs-2017) ) but feel that the rework on the points above will provide a better approach to solve the issue. 
1. Some more optimisations
